src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs

```diff
@@ -77,6 +77,82 @@ internal static ParseResult ParseScript(string script, ILanguageSettings loc = n
             return new ParseResult(parsetree, errors, errors?.Any() ?? false, parser._comments, parser._before, parser._after);
         }
 
+        public static ParseFormulasResult ParseFormulasScript(string script, ILanguageSettings loc = null, Flags flags = Flags.None)
+        {
+            Contracts.AssertValue(script);
+            Contracts.AssertValueOrNull(loc);
+
+            Token[] formulaTokens = TokenizeScript(script, loc, flags);
+            TexlParser parser = new TexlParser(formulaTokens, flags);
+            List<TexlError> errors = null;
+            Dictionary<DName, List<Token>> namedFormulas = parser.ParseFormulas(ref errors);
+
+            // Transform back into script for expression parsing
+            Dictionary<DName, string> namedFormulasScript = new Dictionary<DName, string>();
+            foreach (KeyValuePair<DName,List<Token>> formula in namedFormulas)
+            {
+                var thisExpression = TexlLexer.NewInstance(loc).GetMinifiedScript(script, formula.Value);
+                namedFormulasScript.Add(formula.Key, thisExpression);
+            }
+
+            return new ParseFormulasResult(namedFormulasScript, errors, errors?.Any() ?? false);
+
+        }
+
+        private Dictionary<DName,List<Token>> ParseFormulas(ref List<TexlError> errors)
+        {
+            Dictionary<DName, List<Token>> namedFormulas = new Dictionary<DName, List<Token>>();
+            while (_curs.TokCur.Kind != TokKind.Eof)
+            {
+                // Verify identifier
+                Token thisIdentifier = TokEat(TokKind.Ident);
+                if (thisIdentifier != null)
+                {
+                    // Verify "="
+                    Token thisEq = TokEat(TokKind.Equ);
+                    if (thisEq != null)
+                    {
+                        // Extract expression
+                        List<Token> expression = new List<Token>();
+                        while (_curs.TidCur != TokKind.Semicolon)
+                        {
+                            // Check for a circular reference
+                            if (_curs.TidCur == TokKind.Ident && thisIdentifier.As<IdentToken>().Name == _curs.TokCur.As<IdentToken>().Name)
+                            {
+                                CreateError(_curs.TokCur, TexlStrings.ErrNamedFormula_CircularReference);
```

Comment:
>CreateError(_curs.TokCur, TexlStrings.ErrNamedFormula_CircularReference); [](http://example.com/codeflow?start=32&length=73)

Is this to catch cases like:
   Foo = Foo

I don't think the parser needs to catch that. It's technically a valid parse. Throwing a parse error may prevent the rest of the string from parsing, and we can catch it during binding. 

We do it in binding because that can handle all the cases , like: 
   Foo = Bar
   Bar = Foo

Reply:
Right, yeah. We'll need proper cycle detection in RecalcEngine anyway, and it's part of the Graph in other hosts.

Reply:
Removed